test/core/common/scale_test.cpp

```diff
@@ -0,0 +1,360 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gtest/gtest.h>
+
+#include "common/result.hpp"
+#include "common/scale.hpp"
+
+using namespace kagome;
+using namespace kagome::common;
+using namespace common::scale;
+
+/**
+ * Decode compact tests
+ */
+
+TEST(Scale, compactDecodeZero) {
+  // decode 0
+  std::vector<uint8_t> bytes = {
+      0b00000000,  // 0
+  };
+
+  compact::Decoder decoder{bytes};
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 0);
+  auto value = header.second;
+  ASSERT_EQ(value, 0);  // MIN_UI8
+}
+
+TEST(Scale, compactDecodeOne) {
+  // decode 1
+  std::vector<uint8_t> bytes = {
+      0b00000100,  // 4
+  };
+
+  compact::Decoder decoder{bytes};
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 0);
+  auto value = header.second;
+  ASSERT_EQ(value, 1);  // MIN_UI8
+}
+
+TEST(Scale, compactDecodeMaxUi8) {
+  // decode MAX_UI8 == 63
+  std::vector<uint8_t> bytes = {
+      0b11111100,  // 252
+  };
+
+  compact::Decoder decoder{bytes};
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 0);
+  auto value = header.second;
+  ASSERT_EQ(value, 63);  // MAX_UI8
+}
+
+TEST(Scale, compactDecodeMinUi16) {
+  // decode MIN_UI16 := MAX_UI8+1 == 64
+  std::vector<uint8_t> bytes = {
+      0b00000001,  // 1
+      0b00000001   // 1
+  };
+
+  compact::Decoder decoder{bytes};
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 1);
+  auto value = header.second;
+  ASSERT_EQ(value, 64);  // MAX_UI8 + 1
+}
+
+TEST(Scale, compactDecodeMaxUi16) {
+  // decode MAX_UI16 == 2^14 - 1 == 16383
+  std::vector<uint8_t> bytes = {
+      0b11111101,  // 253
+      0b11111111   // 255
+  };
+
+  compact::Decoder decoder{bytes};
+  auto optionalValue = decoder.decodeInteger();
+  ASSERT_EQ(optionalValue.has_value(), true);
+  auto header = optionalValue.value();
+  auto flag = header.first;
+  ASSERT_EQ(flag, 1);
+  auto value = header.second;
+  ASSERT_EQ(value, 16383);  // MAX_UI16
+}
+
+TEST(Scale, compactDecodeMinUi32) {
```

Comment:
Prefer parametrized tests, so the output is much more useful and code is easier to grasp
http://anadoxin.org/blog/type-parametrized-unit-testing-in-gtest.html

Reply:
I'm going to delve into this article later...